src/framework/mocha/Utilities/MOJSBlock.m

```diff
@@ -113,14 +113,18 @@ static JSValueRef jsInvoke(MOJavaScriptObject* function, NSMethodSignature* sign
         }
     }
     
-    JSContextRef ctx = [function JSContext];
-    Mocha *runtime = [Mocha runtimeWithContext:ctx];
-    
-    JSValueRef value = [runtime callJSFunction:[function JSObject] withArgumentsInArray:arguments];
+    __block JSValueRef value = NULL;
     
-    if (value == NULL) {
-        [NSException raise:@"CallbackError" format:@"Error while calling the callback"];
-    }
+    dispatch_sync(dispatch_get_main_queue(), ^{
```

Comment:
`dispatch_sync` is always a wee bit risky. Are you confident you won't get any deadlocks?

Reply:
I need `dispatch_sync` since I need to return value synchronously. Would there be any other way to do it?

Reply:
Probably. Just want to be sure you've thought about it :-)

Reply:
Well I'm not really confident cause I don't really work with thread usually but I think it's the only solution 😄

Reply:
I realised over the weekend that that a better way to get at the issue is:

Can you guarantee this method is always called **off** of the main thread?

Because if it's ever called on the main thread, it will deadlock. The main thread will be stuck waiting for the main queue to become available to synchronously execute your block.

Reply:
No I can't guarantee that. So do I need to check if we are on the main thread first, then branch?